tests/e2e/improved_csi_idempotency.go

```diff
@@ -0,0 +1,519 @@
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		ginkgo.By(fmt.Sprintf("Sleeping for %v seconds to allow full sync finish", fullSyncWaitTime))
+		time.Sleep(time.Duration(fullSyncWaitTime) * time.Second)
+
+	} else {
+		ginkgo.By(fmt.Sprintf("Stopping %v on the vCenter host", serviceName))
+		vcAddress := e2eVSphere.Config.Global.VCenterHostname + ":" + sshdPort
+		err = invokeVCenterServiceControl(stopOperation, serviceName, vcAddress)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		isServiceStopped = true
+		err = waitVCenterServiceToBeInState(serviceName, vcAddress, "Stopped")
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		defer func() {
+			if isServiceStopped {
+				ginkgo.By(fmt.Sprintf("Starting %v on the vCenter host", serviceName))
+				err = invokeVCenterServiceControl(startOperation, serviceName, vcAddress)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				err = waitVCenterServiceToBeInState(serviceName, vcAddress, "Running")
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			}
+		}()
+
+		ginkgo.By(fmt.Sprintf("Sleeping for %v seconds to allow %v to completely shutdown",
+			vsanHealthServiceWaitTime, serviceName))
+		time.Sleep(time.Duration(vsanHealthServiceWaitTime) * time.Second)
+
+		ginkgo.By(fmt.Sprintln("Sleeping for 5+1 min for default provisioner timeout"))
+		time.Sleep(pollTimeoutSixMin)
+
+		ginkgo.By(fmt.Sprintf("Starting %v on the vCenter host", serviceName))
+		err = invokeVCenterServiceControl(startOperation, serviceName, vcAddress)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		isServiceStopped = false
+		err = waitVCenterServiceToBeInState(serviceName, vcAddress, "Running")
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		//TODO: To replace the sleep with polling
+		//Adding additional buffer sleep time for CSI to issue new events on failures
+		ginkgo.By(fmt.Sprintln("Sleeping for 5 buffer min"))
+		time.Sleep(pollTimeout)
+	}
+
+	//After reboot
+	bootstrap()
+
+	ginkgo.By("Waiting for all claims to be in bound state")
+	persistentvolumes, err = fpv.WaitForPVClaimBoundPhase(client, pvclaims, framework.ClaimProvisionTimeout)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+	defer func() {
+		for _, claim := range pvclaims {
+			err := fpv.DeletePersistentVolumeClaim(client, claim.Name, namespace)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}
+		ginkgo.By("Verify PVs, volumes are deleted from CNS")
+		for _, pv := range persistentvolumes {
+			err := fpv.WaitForPersistentVolumeDeleted(client, pv.Name, framework.Poll, framework.PodDeleteTimeout)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			volumeID := pv.Spec.CSI.VolumeHandle
+			if guestCluster {
+				volumeID = getVolumeIDFromSupervisorCluster(volumeID)
+			}
+			err = e2eVSphere.waitForCNSVolumeToBeDeleted(volumeID)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred(),
+				fmt.Sprintf("Volume: %s should not be present in the CNS after it is deleted from kubernetes", volumeID))
+		}
```

Comment:
where is the check for orphan volumes?

Reply:
Let me address that in the follow-up PR - resolving this for now.